### ISSUE TITLE
Why is YOLOv8_ OBB not very good at square rotating detection?

### ISSUE THREAD

**Issue by zyiiiiiii** [ultralytics/ultralytics#11676]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi All, why is YOLOv8_ OBB not very good at square rotating detection? I have tried many times, the model still can not detect the square rotational objects, like images as follow. But for the square horizontal objects, rectangle horizontal objects and rectangle rotational objects, the model can detect well. It looked like that the model have some challenges to detect square rotational objects.

I look through all of the issues, this issue https://github.com/ultralytics/ultralytics/issues/9394 had a similar problem, but I still want to get more help to fix this problem. Thanks in advance. 

### Additional

From the image we can see that for most of square rotational objects the output bounding boxes always looked like horizontal not rotational.
![val_batch0_pred](https://github.com/ultralytics/ultralytics/assets/166489453/9c4d29eb-6111-4e63-b88f-2f09e2217139)

**Comment by glenn-jocher** [comment]
Hello! It seems like the issue with detecting square rotational objects may stem from the inherent challenges that square objects pose when rotated, as their symmetry can make orientation ambiguous to the model. One potential way to enhance detection capabilities for square rotational objects is to consider augmenting your dataset with more varied examples of these objects in multiple orientations and contexts. Also, if not already done, experimenting with adjustments to the detection thresholds and the anchor settings might prove beneficial.

Here's a basic example of how you might experiment with hyperparameters to adjust the detection settings:

```python
# Example using Python
from ultralytics import YOLO

# Load your model
model = YOLO('path/to/your/model.pt')

# Adjust confidence and intersection over union thresholds
results = model.predict('path/to/image.jpg', conf=0.4, iou=0.6)

# Explore results
print(results.pandas().xyxy[0])  # Print predictions in bounding box format
```

If further tuning does not yield improvements, consider revisiting the data labeling to ensure rotational variances are accurately captured and represented. If you continue to experience difficulties, feel free to share more details or reach out for more specific guidance! 🚀

**Comment by Wangfeng2394** [comment]
I have also encountered such a problem,yolov80bb seems insensitive to angles。

**Comment by zyiiiiiii** [comment]
> I have also encountered such a problem,yolov80bb seems insensitive to angles。
@Wangfeng2394 
Especially for square rotational objects, right? In my case, the model can not detect the square rotational objects well, but for the rectangle rotational objects, it did a good job.

**Comment by Wangfeng2394** [comment]
> > I have also encountered such a problem,yolov80bb seems insensitive to angles。
> > @Wangfeng2394
> > Especially for square rotational objects, right? In my case, the model can not detect the square rotational objects well, but for the rectangle rotational objects, it did a good job.

I you are chinese，give me your  WeChat(微信) ID

**Comment by Wangfeng2394** [comment]
> > > I have also encountered such a problem,yolov80bb seems insensitive to angles。
> > > @Wangfeng2394
> > > Especially for square rotational objects, right? In my case, the model can not detect the square rotational objects well, but for the rectangle rotational objects, it did a good job.
> 
> If you are chinese，give me your WeChat(微信) ID

**Comment by zyiiiiiii** [comment]
> > > > I have also encountered such a problem,yolov80bb seems insensitive to angles。
> > > > @Wangfeng2394
> > > > Especially for square rotational objects, right? In my case, the model can not detect the square rotational objects well, but for the rectangle rotational objects, it did a good job.
> > 
> > 
> > If you are chinese，give me your WeChat(微信) ID
@Wangfeng2394 
Yeah, sure, here is my WeChat jojo_1357.

**Comment by Wangfeng2394** [comment]
> @Wangfeng2394 Hi, I have encountered the same issue in this situation. Have you found a solution to this problem? thx

qq  id or wechat id？

**Comment by glenn-jocher** [comment]
Hello! I'm glad to see the community coming together to solve these issues. 🚀 However, I recommend discussing solutions here or in officially supported channels like GitHub Discussions to benefit everyone. If you have made any progress or experiments you can share (like code adjustments or dataset augmentations), please do share them here! This way, all users facing similar challenges can benefit from your findings. Thanks!

**Comment by Leo5050xvjf** [comment]
> > @Wangfeng2394 Hi, I have encountered the same issue in this situation. Have you found a solution to this problem? thx
> 
> qq id or wechat id？

WeChat ID: wxid_rfrctjb6z0fn12
thanks!

**Comment by Sldora** [comment]
got same issue as well, I guess that's due to the limitation of prob-IoU, change loss function to MSE would get better(at least the bounding box did rotate) but not sure it's precision good enough

**Comment by glenn-jocher** [comment]
Hello! It's great to see you experimenting with different loss functions to tackle the rotation sensitivity issue. Switching to MSE might indeed help with the rotation of bounding boxes due to its different approach to error minimization. If you're seeing improvements in rotation but are still concerned about precision, you might consider combining MSE with another metric that focuses more on classification accuracy. Keep experimenting and let us know how it goes! 😊 If you have any specific results or code snippets from your experiments, sharing them could be very helpful to the community. Thanks for contributing to the discussion!

**Comment by zyiiiiiii** [comment]
> got same issue as well, I guess that's due to the limitation of prob-IoU, change loss function to MSE would get better(at least the bounding box did rotate) but not sure it's precision good enough

@Sldora Yeah, you are right, I also tried to change loss function to MSE, good news is the square rotational objects can detect, but at the cost of losing the accuracy of horizontal objects, for now I keep training and experimenting to make a balance of this.

By the way, do you have any new progress of this?

**Comment by dlwforever** [comment]
@zyiiiiiii Hi, I am struggling on the OBB detection for square objects. Do you have any suggesstions about it? May be we can connect on wechat.

**Comment by zyiiiiiii** [comment]
> @zyiiiiiii Hi, I am struggling on the OBB detection for square objects. Do you have any suggesstions about it? May be we can connect on wechat.
@dlwforever Sure, if you want, please give me your WeChat.

**Comment by dlwforever** [comment]
Thanks for your help. My WeChat ID is HUGD2583. Hopes you all well.&nbsp;


就是个昵称
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人: ***@***.***&gt;; 
发送时间: 2024年6月24日(星期一) 上午8:54
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [ultralytics/ultralytics] Why is YOLOv8_ OBB not very good at square rotating detection? (Issue #11676)














  
@zyiiiiiii Hi, I am struggling on the OBB detection for square objects. Do you have any suggesstions about it? May be we can connect on wechat.
 @dlwforever Sure, if you want, please give me your WeChat.

**Comment by dlwforever** [comment]
OK, My WeChat ID is HUGD2583.

就是个昵称
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "ultralytics/ultralytics"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年6月24日(星期一) 上午8:54
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [ultralytics/ultralytics] Why is YOLOv8_ OBB not very good at square rotating detection? (Issue #11676)





  
@zyiiiiiii Hi, I am struggling on the OBB detection for square objects. Do you have any suggesstions about it? May be we can connect on wechat.
 @dlwforever Sure, if you want, please give me your WeChat.

**Comment by Petros626** [comment]
Note for me

**Comment by glenn-jocher** [comment]
Hello! As mentioned earlier in this thread, please keep discussions here on GitHub so the entire community can benefit from solutions and insights. For the square OBB detection challenges you're all experiencing, I'd recommend focusing on dataset augmentation with more varied square object rotations and experimenting with different confidence thresholds. The community insights about loss functions from @Sldora are worth exploring further too.

**Comment by AUTOMATIC2222** [comment]
> Hello! As mentioned earlier in this thread, please keep discussions here on GitHub so the entire community can benefit from solutions and insights. For the square OBB detection challenges you're all experiencing, I'd recommend focusing on dataset augmentation with more varied square object rotations and experimenting with different confidence thresholds. The community insights about loss functions from [@Sldora](https://github.com/Sldora) are worth exploring further too.

It is no use at all. It just matters loss function, probiou is not consider square object, and that‘s it.

**Comment by glenn-jocher** [comment]
Thank you for the direct feedback on the loss function issue. You're right that if ProbIoU inherently struggles with square object orientation, dataset augmentation alone won't solve the fundamental problem. As @Sldora mentioned earlier, experimenting with MSE loss showed better rotation handling for square objects. Have you tried implementing alternative loss functions, or would you be interested in sharing specific modifications to the ProbIoU implementation that could better handle square geometries?
